helix-term/src/commands.rs

```diff
@@ -2497,25 +2493,37 @@ enum Modified {
 async fn make_format_callback(
     doc_id: DocumentId,
     doc_version: i32,
-    modified: Modified,
     format: impl Future<Output = Result<Transaction, FormatterError>> + Send + 'static,
+    write: Option<(Option<PathBuf>, bool)>,
 ) -> anyhow::Result<job::Callback> {
     let format = format.await?;
-    let call: job::Callback = Box::new(move |editor, _compositor| {
+    let call: job::Callback = Callback::Editor(Box::new(move |editor| {
         let view_id = view!(editor).id;
+        let loader = editor.syn_loader.clone();
+
         if let Some(doc) = editor.document_mut(doc_id) {
             if doc.version() == doc_version {
                 doc.apply(&format, view_id);
                 doc.append_changes_to_history(view_id);
                 doc.detect_indent_and_line_ending();
-                if let Modified::SetUnmodified = modified {
-                    doc.reset_modified();
+
+                if let Some((path, force)) = write {
+                    let refresh_lang = path.is_some();
+
+                    if let Err(err) = doc.save(path, force) {
+                        editor.set_error(format!("Error saving: {}", err));
+                    } else if refresh_lang {
+                        let id = doc.id();
+                        doc.detect_language(loader);
+                        let _ = editor.refresh_language_server(id);
```

Comment:
Does this still run even if there was no format callback? We want to run language detection + refresh if a new file is saved with a name. `:w test.rs` should start highlighting as rust

Reply:
Might be better to put this into application.rs, right after `set_path`

Reply:
No, this code is inside the callback, but that is actually a good point: the write could fail, so doing this here is a bit premature. I agree, I'll move this into the document saved event handler.